tests/request-response/request_response_protocol_adapter_tests.c

```diff
@@ -0,0 +1,1011 @@
+    s_request_response_mqtt5_protocol_adapter_subscribe_success_fn)
+
+static int s_request_response_mqtt5_protocol_adapter_subscribe_failure_timeout_fn(
+    struct aws_allocator *allocator,
+    void *ctx) {
+    (void)ctx;
+
+    ASSERT_SUCCESS(s_do_request_response_mqtt5_protocol_adapter_subscribe_test(allocator, PAOTT_FAILURE_TIMEOUT));
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(
+    request_response_mqtt5_protocol_adapter_subscribe_failure_timeout,
+    s_request_response_mqtt5_protocol_adapter_subscribe_failure_timeout_fn)
+
+static int s_request_response_mqtt5_protocol_adapter_subscribe_failure_reason_code_fn(
+    struct aws_allocator *allocator,
+    void *ctx) {
+    (void)ctx;
+
+    ASSERT_SUCCESS(s_do_request_response_mqtt5_protocol_adapter_subscribe_test(allocator, PAOTT_FAILURE_REASON_CODE));
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(
+    request_response_mqtt5_protocol_adapter_subscribe_failure_reason_code,
+    s_request_response_mqtt5_protocol_adapter_subscribe_failure_reason_code_fn)
+
+static int s_request_response_mqtt5_protocol_adapter_subscribe_failure_error_code_fn(
+    struct aws_allocator *allocator,
+    void *ctx) {
+    (void)ctx;
+
+    ASSERT_SUCCESS(s_do_request_response_mqtt5_protocol_adapter_subscribe_test(allocator, PAOTT_FAILURE_ERROR_CODE));
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(
+    request_response_mqtt5_protocol_adapter_subscribe_failure_error_code,
+    s_request_response_mqtt5_protocol_adapter_subscribe_failure_error_code_fn)
+
+static enum aws_mqtt5_unsuback_reason_code s_failed_unsuback_reason_codes[] = {
+    AWS_MQTT5_UARC_NOT_AUTHORIZED,
+};
+
+static int s_aws_mqtt5_server_send_failed_unsuback_on_unsubscribe(
+    void *packet,
+    struct aws_mqtt5_server_mock_connection_context *connection,
+    void *user_data) {
+    (void)packet;
+    (void)user_data;
+
+    struct aws_mqtt5_packet_subscribe_view *unsubscribe_view = packet;
+
+    struct aws_mqtt5_packet_unsuback_view unsuback_view = {
+        .packet_id = unsubscribe_view->packet_id,
+        .reason_code_count = AWS_ARRAY_SIZE(s_failed_unsuback_reason_codes),
+        .reason_codes = s_failed_unsuback_reason_codes,
+    };
+
+    return aws_mqtt5_mock_server_send_packet(connection, AWS_MQTT5_PT_UNSUBACK, &unsuback_view);
+}
+
+static int s_do_request_response_mqtt5_protocol_adapter_unsubscribe_test(
+    struct aws_allocator *allocator,
+    enum protocol_adapter_operation_test_type test_type) {
+    aws_mqtt_library_init(allocator);
+
+    struct mqtt5_client_test_options test_options;
+    aws_mqtt5_client_test_init_default_options(&test_options);
+
+    if (test_type == PAOTT_SUCCESS) {
+        test_options.server_function_table.packet_handlers[AWS_MQTT5_PT_UNSUBSCRIBE] =
+            aws_mqtt5_mock_server_handle_unsubscribe_unsuback_success;
+    } else if (test_type == PAOTT_FAILURE_REASON_CODE) {
+        test_options.server_function_table.packet_handlers[AWS_MQTT5_PT_UNSUBSCRIBE] =
```

Comment:
Here and for subscribe and publish tests, the function should be set to a separate "timeout" function instead of the failed. On timeout, the client doesn't receive the same type of packet it would from the server and instead fails with `AWS_ERROR_MQTT_TIMEOUT` error. This would properly simulate what the adapter would be receiving and handling. If we don't do that, we probably don't need to have separate timeout failure tests because functionally they aren't taking any different path than failure error code and failure reason code.

Reply:
I don't understand what you're saying here.  When it's a timeout test there's no packet handler so the operation times out.

Reply:
Ah, my mistake. I thought `s_aws_mqtt5_server_send_failed_unsuback_on_unsubscribe` was being assigned for all non successful PAOTT. The timeout tests are correctly checking against timeout.